src/content/blog/hybrid-rendering.mdx

```diff
@@ -0,0 +1,114 @@
+
+<BlogContentImage
+    src={staticBuild}
+    alt="Flow diagram of Astro's static build process"
+/>
+
+If your configured [`output`](https://docs.astro.build/en/reference/configuration-reference/#output) is `server`, Astro passes the server-side JavaScript to an [`adapter`](https://docs.astro.build/en/reference/configuration-reference/#adapter) for further processing.
+Adapters ensure that your server-side JavaScript is compatible with a particular hosting provider's JavaScript runtime. Note that in this case, the final output is not a set of `.html` files, but the JavaScript code necessary to render HTML.
+
+<BlogContentImage
+    src={serverBuild}
+    alt="Flow diagram of Astro's server build process"
+/>
+
+## How does hybrid rendering work?
+
+In Astro 2.0, we've revamped the `astro build` pipeline to handle hybrid rendering.
+During the initial bundling process, a new static analysis step determines which pages should be prerendered.
+This allows us to split your routes into seperate chunks, depending on _when_ they should be rendered.
+
+Much like the original `static` process, the prerendered chunk is executed and the output is written to `.html` files. The chunk is then discarded.
+
+Much like the original `server` process, the server chunk is passed to an [`adapter`](https://docs.astro.build/en/reference/configuration-reference/#adapter) for further processing. It will ultimately be deployed as a Serverless or Edge function, depending on your adapter.
+
+<BlogContentImage
+    src={hybridBuild}
+    alt="Flow diagram of Astro's new hybrid build process"
+/>
+
+## Why use static analysis?
+
+Static analysis is a technique used to analyze your source code without actually executing it.
+Astro relies on this approach in many places, but it is especially critical during the build process.
+For performance reasons, Astro must be able to categorize your pages without executing their source code.
+For smaller projects, execution might not be a concern, but at scale it quickly becomes a bottleneck.
+
+Hybrid rendering was specifically designed with static analysis in mind, though we did [entertain alternative APIs](https://github.com/withastro/roadmap/blob/main/proposals/0029-prerender-api.md#alternatives).
+Astro's build process is able to check for the presence of an `export const prerender = true` statement to determine which pages should be prerendered.
+To do this efficiently, we rely on the wonderful [`es-module-lexer`](https://github.com/guybedford/es-module-lexer) library, which is also used internally in Node.
+
+```astro
+---
+// This route should be generated at build time!
+export const prerender = true
+
+const text = await fetch('https://example.com/').then((res) => res.text())
+---
+
+<article set:html={text} />
+```
+
+## Use cases for hybrid rendering
```

Comment:
Subheadings are rarely useful in a blog post IMO, since someone is reading it top-to-bottom they aren't organizing the information in the same way as docs.

I would push away from the "sub-headings" here and go back to the three `##` headings that all start with `## Use case:` or `## Case Study:`

Reply:
It's okay if the sections are short, although the second one "Add an API" could use a second paragraph, even if it  a bit of filler.